apps/emqx_rule_engine/src/emqx_rule_engine.erl

```diff
@@ -459,26 +458,27 @@ parse_and_insert(Params = #{id := RuleId, sql := Sql, actions := Actions}, Creat
                 conditions => emqx_rule_sqlparser:select_where(Select)
                 %% -- calculated fields end
             },
-            ok = insert_rule(Rule),
+            ok = save_rule(Rule),
+            ok = load_rule_hook_and_metrics(Rule),
             {ok, Rule};
         {error, Reason} ->
             {error, Reason}
     end.
 
-do_insert_rule(#{id := Id} = Rule) ->
+save_rule(#{id := RuleId} = Rule) ->
+    persistent_term:put(?RULE_ENGINE_PT_KEY, (get_rules_map())#{RuleId => Rule}).
```

Comment:
Yes, it is.

> If the deleted term (or any part of it) is **still used by a process**, that process will undergo a major (fullsweep) garbage collection and copy the term into the process. However, only up to two processes at a time will be scheduled for such garbage collection.
>  -- [persistent_term](https://www.erlang.org/doc/man/persistent_term.html)

`still used by a process` strictly speaking, when updating this persistent term, only the processes currently holding it should be affected. This may just be a subset of all processes involved? although it could potentially make up a significant portion.

I don't know if my understanding is correct. and I'm not quite sure how much worse it will make things? Especially with millions of connection processes and hundreds of rules. 😿 

BTW, our configuration system currently also uses persistent_terms to implement it, which sounds like it will have the same problem as well. 🤔

Reply:
Yes. but overall the cpu usage dropped by around about 48% with continuous pub-sub.

100 Rules(Replish)
100 clients pub to `t/1` \* 2 msg/(s\*perclient)
5 clients sub to `a/1`~`j/1` (total 5*10)

| TYPE                    | CPU         | MEM(REG) | MEM(DATA) | MEM(VIRT) |
|-------------------------|-------------|----------|-----------|-----------|
| ets                     | 720% ~ 800% | 530M     | 830M      | 5380M     |
| persistent\_term(lists) | 380% ~ 460% | 420M     | 720M      | 5450M     |
| persistent\_term(map)   | 350% ~ 460% | 410M     | 710M      | 5440M     |

Reply:
> Especially with millions of connection processes

If, for example, some process hit `emqx_rule_engine:on_message_publish/2` function it will likely need full-sweep GC, even if no rule has been triggered.

> BTW, our configuration system currently also uses persistent_terms to implement it, which sounds like it will have the same problem as well.

Yes, but likely not all the time. If you get immediate term with `emqx_config:get/2` (i.e. atom, number, probably long enough binary), then full-sweep GC might be avoided (though I'm not entirely sure about that).

Reply:
> but overall the cpu usage dropped by around about 48% with continuous pub-sub.

Nice results indeed!

But just to be clear, my point is: my hunch is that just adding an index for the `from` field to the same ETS table and using it in some hot paths (`get_rules_for_topic/0`, `get_rules_with_same_event/0`) will give us comparable or even better results, while keeping us away from all the nasty side-effects of persistent terms and associated complexity (i.e. separate lifecycle and potential inconsistencies across application restarts, expensive GCs, etc.). Again, this is just a hunch, take it with a grain of salt, but I would experiment with that approach before resorting to persistent terms.

Reply:
Disregard my previous comment, I've realized just now that indexing `from` field is not that simple since we need to implement `emqx_topic:match_any/2` logic on top of it (which is a quite interesting challenge by itself to be honest).

Reply:
True. If I remember correctly, the process of retaining messages implemented indexing the `Topic Name` field to search a large number of messages. This differs slightly from the rule engine. The retainer utilizes Topic Filter to query the Topic Names in ETS, whereas the rule engine requires using a Topic Name to match the Topic Filters.

In future developments, it may be worth considering creating generic libraries for this purpose 🙊

Anyway, thank you for your suggestions and join the discussion 😄